tempesta_fw/sock_srv.c

```diff
@@ -416,13 +417,13 @@ tfw_sock_srv_disconnect(TfwConn *conn)
 	 * resources. Otherwise, use synchronous closing to ensure that
 	 * the job is enqueued.
 	 */
-	if (atomic_read(&conn->refcnt) == TFW_CONN_DEATHCNT) {
-		if (conn->sk)
-			ret = ss_close_sync(conn->sk, true);
+	if (atomic_read(&conn->refcnt) == TFW_CONN_DEATHCNT)
 		tfw_connection_release(conn);
-	}
-	else {
-		ret = ss_close_sync(conn->sk, true);
+	if (sk) {
+		sock_hold(sk);
+		tfw_connection_unlink_to_sk(conn);
```

Comment:
Again, I'm not sure about correctness of the change. We had many crashes when various contexts resolve already dead `conn->sk`, so now we call `tfw_connection_unlink_to_sk()` on connection destructor only, when `tfw_connection_put()` ensures that there are no more users of the connection. Just removing the link could lead to inconsistencies.